### ISSUE TITLE
Jimp Browser Support

### ISSUE THREAD

**Issue by marvinruder** [jimp-dev/jimp#1194]
**Is your feature request related to a problem? Please describe.**
At the moment, some problems arise when attempting to use Jimp in React (similar issues were reported from users of other tools, e.g. React Native). The following are known so far:

- [ ] `Jimp.read` is undefined / not a function – can be avoided by using Typescript’s `useModuleInterop: true`, no fix yet for plain Javascript. Reproducable in [jimp-dev/jimp-react](https://github.com/jimp-dev/jimp-react). Might be related to #1191, #1178, #761.
- [ ] `process is not defined` in browser console, caused by `process.env.__TESTING_MKDIRP_PLATFORM__` in dependency `mkdirp`
- [ ] When using the `jimp/es` import, accessing some built-in Node APIs like `fs`, `path` etc. is attempted and the build fails (example: `Can't resolve 'fs' in '…/node_modules/@jimp/core/es'`). Reproducable in [jimp-dev/jimp-react](https://github.com/jimp-dev/jimp-react)

**Describe the solution you'd like**
The issues described above are resolved and Jimp can be used in frontend tools like React using `import Jimp from "jimp[/es]";` without a problem. 

**Describe alternatives you've considered**
Using the browser version of Jimp via `require("jimp/browser/lib/jimp.js"); const { Jimp } = window;` is a possible workaround, but that version is aimed more at using Jimp in plain HTML/JS than by bundling tools such as Webpack, Rollup etc., which should be able to handle `jimp` or `jimp/es` properly.

**Comment by sergeu90** [comment]
![image](https://user-images.githubusercontent.com/7217517/217366849-2fe9a148-e80f-4e39-855b-bb0036e0d0c8.png)

**Comment by crutchcorn** [comment]
It might be worthwhile to explore an option similar to how @wmhilton has done so with `isomorphic-git` using FS as a property to pass in:

https://isomorphic-git.org/docs/en/fs#docsNav

@wmhilton has even created a browser-first FS package called "LighningFS" that might be a good package to point to for official browser support, alongside an example repo that integrates it:

https://github.com/isomorphic-git/lightning-fs

**Comment by crutchcorn** [comment]
Also worth mentioning - @marvinruder you may want to look into the work being done by @danielholmes :

https://github.com/danielholmes/jimp-vite

**Comment by suneox** [comment]
It can work well on [example](https://github.com/jimp-dev/jimp/tree/main/packages/jimp/browser) so I have resolved the workaround by loading jimp.js on index.html and defining global type for Jimp variable

**Comment by hipstersmoothie** [comment]
Yeah that one works but we're talking about esm usage rather than bundling. Bundling means we can't tree shake things. Right now our esm breaks browsers

**Comment by gabycush** [comment]
> When using the jimp/es import, accessing some built-in Node APIs like fs, path etc. is attempted and the build fails (example: Can't resolve 'fs' in '…/node_modules/@jimp/core/es'). Reproducable in [jimp-dev/jimp-react](https://github.com/jimp-dev/jimp-react)

I'm having this issue too, is a solution in the works or is there a specific suggested workaround?

**Comment by Ronkiro** [comment]
My solution for NextJS was to load jimp and import from window, similar approach from the issue's text but not using the library for the browser (it was giving me the error that `self is not defined`).

Code:
```
import 'jimp';
const { Jimp } = window as any;
```

Hope it helps anyone

**Comment by angello0926** [comment]
> My solution for NextJS was to load jimp and import from window, similar approach from the issue's text but not using the library for the browser (it was giving me the error that `self is not defined`).
> 
> Code:
> 
> ```
> import 'jimp';
> const { Jimp } = window as any;
> ```
> 
> Hope it helps anyone

This really helps me to get through this issue for now! Struggled for a few hours, thanks!

**Comment by MBrunoS** [comment]
I tried all the suggestions, but still couldn't make it work... I'm trying to use Jimp in a Web Worker, with Vite. So, I'm not being able to use the lib.
I'd even help with this bug, but I don't know where to start right now.

**Comment by xenova** [comment]
I got Jimp working in the browser/webworker/... with this ugly hack:

ES6:
```js
// For some reason, Jimp attaches to self, even in Node.
// https://github.com/jimp-dev/jimp/issues/466
import _Jimp from 'jimp';

// @ts-ignore
const Jimp = (typeof self !== 'undefined') ? (self.Jimp || _Jimp) : _Jimp;
```

CommonJS:
```js
// For some reason, Jimp attaches to self, even in Node.
// https://github.com/jimp-dev/jimp/issues/466
const _Jimp = require('jimp');

// @ts-ignore
const Jimp = (typeof self !== 'undefined') ? (self.Jimp || _Jimp) : _Jimp;
```

**Comment by sexnine** [comment]
@xenova I don't know why this works, but it works.

I've been struggling to get Jimp working on Cloudflare Workers (using SvelteKit) for wayyy too long and you're solution just works.

One thing I noticed using your solution is instantiating a new Jimp instance I have to do `new Jimp.default(...)` on Node.js but on Cloudflare Workers `new Jimp(...)` is the only way that works.

I am extremely grateful that you shared your solution, so thanks <3.

**Comment by xenova** [comment]
> I don't know why this works, but it works.

That's exactly what I thought when I got it working 🤣 ... 

> One thing I noticed using your solution is instantiating a new Jimp instance I have to do new Jimp.default(...) on Node.js but on Cloudflare Workers new Jimp(...) is the only way that works.

Interesting. I haven't had an issue with creating new Jimp instances with my node.js version... but it's good to know you found a solution!

> I am extremely grateful that you shared your solution, so thanks <3.

Happy to help!

**Comment by LetMeDream** [comment]
> 



> > My solution for NextJS was to load jimp and import from window, similar approach from the issue's text but not using the library for the browser (it was giving me the error that `self is not defined`).
> > Code:
> > ```
> > import 'jimp';
> > const { Jimp } = window as any;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hope it helps anyone
> 
> This really helps me to get through this issue for now! Struggled for a few hours, thanks!



> My solution for NextJS was to load jimp and import from window, similar approach from the issue's text but not using the library for the browser (it was giving me the error that `self is not defined`).
> 
> Code:
> 
> ```
> import 'jimp';
> const { Jimp } = window as any;
> ```
> 
> Hope it helps anyone

Ok but, where is that code supposed to go inside the next Js project? In the component that's gonna use Jimp?

**Comment by LetMeDream** [comment]
> My solution for NextJS was to load jimp and import from window, similar approach from the issue's text but not using the library for the browser (it was giving me the error that `self is not defined`).
> 
> Code:
> 
> ```
> import 'jimp';
> const { Jimp } = window as any;
> ```
> 
> Hope it helps anyone

Ok but, where is that code supposed to go inside the next Js project? In the component that's gonna use Jimp?

**Comment by LetMeDream** [comment]
> I got Jimp working in the browser/webworker/node/... with this ugly hack:
> 
> ES6:
> 
> ```js
> // For some reason, Jimp attaches to self, even in Node.
> // https://github.com/jimp-dev/jimp/issues/466
> import * as _Jimp from 'jimp';
> 
> // @ts-ignore
> const Jimp = (typeof self !== 'undefined') ? (self.Jimp || _Jimp) : _Jimp;
> ```
> 
> CommonJS:
> 
> ```js
> // For some reason, Jimp attaches to self, even in Node.
> // https://github.com/jimp-dev/jimp/issues/466
> const _Jimp = require('jimp');
> 
> // @ts-ignore
> const Jimp = (typeof self !== 'undefined') ? (self.Jimp || _Jimp) : _Jimp;
> ```

This works in Next JS and apps created using CRA. Not in Vite, tho.

**Comment by patrickelectric** [comment]
No luck when using it in a webpage in a typescript project:
```
[vue warn]: Error in v-on handler (Promise/async): "TypeError: jimp__WEBPACK_IMPORTED_MODULE_7___default(...) is not a function"

found in

---> <VBtn>
       <GenericViewer> at src/components/vehiclesetup/viewers/GenericViewer.vue
         <VehicleViewer> at src/components/vehiclesetup/viewers/VehicleViewer.vue
           <VCard>
             <SetupOverview> at src/components/vehiclesetup/SetupOverview.vue
               <VehicleSetupView> at src/views/VehicleSetupView.vue
                 <VMain>
                   <VApp>
                     <App> at src/App.vue
                       <Root>
```
w

**Comment by Lofavreel** [comment]
> I got Jimp working in the browser/webworker/node/... with this ugly hack:
> 
> ES6:
> 
> ```js
> // For some reason, Jimp attaches to self, even in Node.
> // https://github.com/jimp-dev/jimp/issues/466
> import * as _Jimp from 'jimp';
> 
> // @ts-ignore
> const Jimp = (typeof self !== 'undefined') ? (self.Jimp || _Jimp) : _Jimp;
> ```
> 
> CommonJS:
> 
> ```js
> // For some reason, Jimp attaches to self, even in Node.
> // https://github.com/jimp-dev/jimp/issues/466
> const _Jimp = require('jimp');
> 
> // @ts-ignore
> const Jimp = (typeof self !== 'undefined') ? (self.Jimp || _Jimp) : _Jimp;
> ```

I just did a little digging, because I had the same issue. I found this in the @jimp/core/ index.js (in both es and dist):
This is before the last three lines of code in those files:
`if (process.env.ENVIRONMENT === "BROWSER") {
  // For use in a web browser or web worker
  /* global self */
  let gl;
  if (typeof window !== "undefined" && typeof window === "object") {
    gl = window;
  }
  if (typeof self !== "undefined" && typeof self === "object") {
    gl = self;
  }
  gl.Jimp = Jimp;
  gl.Buffer = Buffer;
}`

I don't know why and I don't think I will look into it, but maybe this will give the right person a starting idea on where to look to resolve this issue.

**Comment by hipstersmoothie** [comment]
Any help from someone in the community would be appreciated. If you want to submit a PR I'd be happy to Shepard that PR. At the time unfortunately I don't see myself doing this soon.

**Comment by jcard0na** [comment]
Just of others usign React, the workaround mentioned on the issue's text worked for me:

```js
export default function App() {
    require("jimp/browser/lib/jimp.js"); 
    const { Jimp } = window
    ...
};
```

this also worked:

```js
import 'jimp';

export default function App() {
    const { Jimp } = window
    ...
};

**Comment by devenkhatri** [comment]
> > My solution for NextJS was to load jimp and import from window, similar approach from the issue's text but not using the library for the browser (it was giving me the error that `self is not defined`).
> > Code:
> > ```
> > import 'jimp';
> > const { Jimp } = window as any;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hope it helps anyone
> 
> This really helps me to get through this issue for now! Struggled for a few hours, thanks!

Thanks for this suggestion. This worked for me.

**Comment by da7a90-backup** [comment]
For Vercel edge the workaround mentioned by @xenova works locally but not on production and even locally you have to set `unstable_allowDynamic` and point it to the library in node_modules but that doesn't seem to work for some reason when trying to deploy to production.

**Comment by sergeu90** [comment]
My solution. Version 0.22.10
```
import * as _Jimp from 'jimp/browser/lib/jimp';
const { Jimp } = window;
```

**Comment by meepen** [comment]
none of these work very well for typescript...

**Comment by Ronkiro** [comment]
> none of these work very well for typescript...

@meepen  check my comment https://github.com/jimp-dev/jimp/issues/1194#issuecomment-1449287374 it may help you

**Comment by sameert89** [comment]
For react native/expo with hermes engine, this actually works, I cannot understand why is it like that, any insigths would be very helpful.

> ```js
> import * as _Jimp from 'jimp';
> const Jimp = (typeof self !== 'undefined') ? (self.Jimp || _Jimp) : _Jimp;
> ```

**Comment by Ronkiro** [comment]
@sameert89  React Native newbie here, but it seems that `self` works like `window` for RN. This snippet checks if it's existent/accessible, if it is then use it, else use the normal variable. It's similar to the solution i pointed https://github.com/jimp-dev/jimp/issues/1194#issuecomment-1449287374 and to the one pointed in the first post

**Comment by NayamAmarshe** [comment]
> none of these work very well for typescript...

Maybe try this:

```jsx
import "jimp";

const ReactComponent = () => {
  const { Jimp } = window as Window & typeof globalThis & {
    Jimp: typeof import("jimp");
  };
...
```

**Comment by tsgouros** [comment]
> For react native/expo with hermes engine, this actually works, I cannot understand why it is like that, any insigths would be very helpful.
> 
> > ```js
> > import * as _Jimp from 'jimp';
> > const Jimp = (typeof self !== 'undefined') ? (self.Jimp || _Jimp) : _Jimp;
> > ```

I was very excited to find this comment, since I am trying to use Jimp in React Native (w/o Expo), but it does not work for me. FWIW, I put a console.log(_Jimp) in between those lines, and it returns this:

```{"default": {}}```

**Comment by neko-para** [comment]
For those want to use jimp in browser with typescript definition, here's a guide.
First, import browser build. As it pollute into window.Jimp, place doesn't matter.
```typescript
import 'jimp/browser/lib/jimp.js'
```
Then, add a dts file
```typescript
declare var Jimp: typeof import('jimp')
```
Just use `Jimp.xxx`.

**Comment by vickyRathee** [comment]
@sexnine Can you please prove the example code for Cloudflare? I can't get it work for some reason.

**Comment by sexnine** [comment]
> @sexnine Can you please prove the example code for Cloudflare? I can't get it work for some reason. 

Are you using Sveltekit on CF workers, or just vanilla workers?  Typescript or JS?

I can _try_ to make a quick demo later today :)

**Comment by vickyRathee** [comment]
Just a vanilla worker using Typescript. I am trying to resize a puppeteer screenshot  and upload to R2 to generate website thumbnail.

This was my code when I was testing with `sharp`

![image](https://github.com/jimp-dev/jimp/assets/6106479/609b617b-46f3-46fb-8104-ce4d7f997db4)

**Comment by trungpd1102** [comment]
> Just a vanilla worker using Typescript. I am trying to resize a puppeteer screenshot and upload to R2 to generate website thumbnail.
> 
> This was my code when I was testing with `sharp`
> 
> ![image](https://private-user-images.githubusercontent.com/6106479/324814466-609b617b-46f3-46fb-8104-ce4d7f997db4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzM3NzgsIm5iZiI6MTcxNTMzMzQ3OCwicGF0aCI6Ii82MTA2NDc5LzMyNDgxNDQ2Ni02MDliNjE3Yi00NmYzLTQ2ZmItODEwNC1jZTRkN2Y5OTdkYjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTBUMDkzMTE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDM5NTdkZDU4ZDFlNGIxZGJlZjNlYjY3Y2RiMTA2NzZjNWJiNWM1Mzc0MzkxMjc3MWI4ZTczNGY3NzRjYWI0YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.XzO_SNcaUWy3eXJfPQLjdwQarfDA4Gn1rXnoNPm2aHc)

Sharp include fs is NodeJs build-in function fs. So it can only run on supported NodeJs.
Unfortunately, Workers does not support all NodeJs function

**Comment by hipstersmoothie** [comment]
:rocket: Issue was released in [`v1.0.2`](https://github.com/jimp-dev/jimp/releases/tag/v1.0.2) :rocket:
